File: dream-server/extensions/services/dashboard-api/routers/workflows.py
"""Workflow management endpoints — n8n integration."""

import json
import logging
import re

import aiohttp
from fastapi import APIRouter, Depends, HTTPException

from config import (
    SERVICES, WORKFLOW_DIR, WORKFLOW_CATALOG_FILE,
    DEFAULT_WORKFLOW_CATALOG, N8N_URL, N8N_API_KEY,
)
from security import verify_api_key

logger = logging.getLogger(__name__)
router = APIRouter(tags=["workflows"])


# --- Helpers ---

def load_workflow_catalog() -> dict:
    """Load workflow catalog from JSON file."""
    if not WORKFLOW_CATALOG_FILE.exists():
        return DEFAULT_WORKFLOW_CATALOG
    try:
        with open(WORKFLOW_CATALOG_FILE) as f:
            data = json.load(f)
        if not isinstance(data, dict):
            logger.warning("Workflow catalog must be a JSON object: %s", WORKFLOW_CATALOG_FILE)
            return DEFAULT_WORKFLOW_CATALOG
        workflows = data.get("workflows", [])
        categories = data.get("categories", {})
        if not isinstance(workflows, list):
            workflows = []
        if not isinstance(categories, dict):
            categories = {}
        return {"workflows": workflows, "categories": categories}
    except (json.JSONDecodeError, OSError, KeyError) as e:
        logger.warning("Failed to load workflow catalog from %s: %s", WORKFLOW_CATALOG_FILE, e)
        return DEFAULT_WORKFLOW_CATALOG


async def get_n8n_workflows() -> list[dict]:
    """Get all workflows from n8n API."""
    try:
        headers = {}
        if N8N_API_KEY:
            headers["X-N8N-API-KEY"] = N8N_API_KEY
        async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=5)) as session:
            async with session.get(f"{N8N_URL}/api/v1/workflows", headers=headers) as resp:
                if resp.status == 200:
                    data = await resp.json()
                    return data.get("data", [])
    except (aiohttp.ClientError, OSError, json.JSONDecodeError) as e:
        logger.warning(f"Failed to fetch workflows from n8n: {e}")
    return []


async def check_workflow_dependencies(deps: list[str], health_cache: dict[str, bool] | None = None) -> dict[str, bool]:
    """Check if required services are running. Uses health_cache to avoid duplicate checks."""
    from helpers import check_service_health

    _DEP_ALIASES = {"ollama": "llama-server"}
    if health_cache is None:
        health_cache = {}
    results = {}
    for dep in deps:
        resolved = _DEP_ALIASES.get(dep, dep)
        if resolved in health_cache:
            results[dep] = health_cache[resolved]
        elif resolved in SERVICES:
            status = await check_service_health(resolved, SERVICES[resolved])
            healthy = status.status == "healthy"
            health_cache[resolved] = healthy
            results[dep] = healthy
        else:
            results[dep] = True
    return results


async def check_n8n_available() -> bool:
    """Check if n8n is responding."""
    try:
        async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=3)) as session:
            async with session.get(f"{N8N_URL}/healthz") as resp:
                return resp.status < 500
    except (aiohttp.ClientError, OSError):
        return False


# --- Endpoints ---

@router.get("/api/workflows")
async def api_workflows(api_key: str = Depends(verify_api_key)):
    """Get workflow catalog with status and dependency info."""
    catalog = load_workflow_catalog()
    n8n_workflows = await get_n8n_workflows()
    n8n_by_name = {w.get("name", "").lower(): w for w in n8n_workflows}

    workflows = []
    health_cache: dict[str, bool] = {}
    for wf in catalog.get("workflows", []):
        wf_name_lower = wf["name"].lower()
        installed = None
        for n8n_name, n8n_wf in n8n_by_name.items():
            if wf_name_lower in n8n_name or n8n_name in wf_name_lower:
                installed = n8n_wf
                break

        dep_status = await check_workflow_dependencies(wf.get("dependencies", []), health_cache)
        all_deps_met = all(dep_status.values())

        executions = 0
        if installed:
            executions = installed.get("statistics", {}).get("executions", {}).get("total", 0)

        workflows.append({
            "id": wf["id"],
            "name": wf["name"],
            "description": wf["description"],
            "icon": wf.get("icon", "Workflow"),
            "category": wf.get("category", "general"),
            "status": "active" if installed and installed.get("active") else ("installed" if installed else "available"),
            "installed": installed is not None,
            "active": installed.get("active", False) if installed else False,
            "n8nId": installed.get("id") if installed else None,
            "dependencies": wf.get("dependencies", []),
            "dependencyStatus": dep_status,
            "allDependenciesMet": all_deps_met,
            "diagram": wf.get("diagram", {}),
            "setupTime": wf.get("setupTime", "~2 min"),
            "executions": executions,
            "featured": wf.get("featured", False)
        })

    return {
        "workflows": workflows,
        "categories": catalog.get("categories", {}),
        "catalogSource": str(WORKFLOW_CATALOG_FILE),
        "workflowDir": str(WORKFLOW_DIR),
        "n8nUrl": N8N_URL,
        "n8nAvailable": len(n8n_workflows) > 0 or await check_n8n_available()
    }


@router.post("/api/workflows/{workflow_id}/enable")
async def enable_workflow(workflow_id: str, api_key: str = Depends(verify_api_key)):
    """Import a workflow template into n8n."""
    if not re.match(r'^[a-zA-Z0-9_-]+$', workflow_id):
        raise HTTPException(status_code=400, detail="Invalid workflow ID format")

    catalog = load_workflow_catalog()
    wf_info = next((wf for wf in catalog.get("workflows", []) if wf["id"] == workflow_id), None)
    if not wf_info:
        raise HTTPException(status_code=404, detail=f"Workflow not found: {workflow_id}")

    dep_status = await check_workflow_dependencies(wf_info.get("dependencies", []))
    missing_deps = [dep for dep, ok in dep_status.items() if not ok]
    if missing_deps:
        raise HTTPException(status_code=400, detail=f"Missing dependencies: {', '.join(missing_deps)}. Enable these services first.")

    workflow_file = WORKFLOW_DIR / wf_info["file"]
    try:
        workflow_file = workflow_file.resolve()
        if not str(workflow_file).startswith(str(WORKFLOW_DIR.resolve())):
            raise HTTPException(status_code=400, detail="Invalid workflow file path")
    except HTTPException:
        raise
    except (OSError, ValueError):
        raise HTTPException(status_code=400, detail="Invalid workflow file path")

    if not workflow_file.exists():
        raise HTTPException(status_code=404, detail=f"Workflow file not found: {wf_info['file']}")

    try:
        with open(workflow_file) as f:
            workflow_data = json.load(f)
    except (OSError, json.JSONDecodeError) as e:
        raise HTTPException(status_code=500, detail=f"Failed to read workflow: {e}")

    try:
        headers = {"Content-Type": "application/json"}
        if N8N_API_KEY:
            headers["X-N8N-API-KEY"] = N8N_API_KEY

        async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=10)) as session:
            async with session.post(f"{N8N_URL}/api/v1/workflows", headers=headers, json=workflow_data) as resp:
                if resp.status in (200, 201):
                    result = await resp.json()
                    n8n_id = result.get("data", {}).get("id")
                    activated = False
                    if n8n_id:
                        async with session.patch(f"{N8N_URL}/api/v1/workflows/{n8n_id}", headers=headers, json={"active": True}) as activate_resp:
                            activated = activate_resp.status == 200
                    return {"status": "success", "workflowId": workflow_id, "n8nId": n8n_id, "activated": activated, "message": f"{wf_info['name']} is now active!"}
                else:
                    error_text = await resp.text()
                    raise HTTPException(status_code=resp.status, detail=f"n8n API error: {error_text}")
    except aiohttp.ClientError as e:
        raise HTTPException(status_code=503, detail=f"Cannot reach n8n: {e}")


@router.delete("/api/workflows/{workflow_id}")
async def disable_workflow(workflow_id: str, api_key: str = Depends(verify_api_key)):
    """Remove a workflow from n8n."""
    n8n_workflows = await get_n8n_workflows()
    catalog = load_workflow_catalog()
    wf_info = next((wf for wf in catalog.get("workflows", []) if wf["id"] == workflow_id), None)
    if not wf_info:
        raise HTTPException(status_code=404, detail=f"Workflow not found: {workflow_id}")

    n8n_wf = None
    wf_name_lower = wf_info["name"].lower()
    for wf in n8n_workflows:
        if wf_name_lower in wf.get("name", "").lower():
            n8n_wf = wf
            break
    if not n8n_wf:
        raise HTTPException(status_code=404, detail="Workflow not installed in n8n")

    try:
        headers = {}
        if N8N_API_KEY:
            headers["X-N8N-API-KEY"] = N8N_API_KEY
        async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=5)) as session:
            async with session.delete(f"{N8N_URL}/api/v1/workflows/{n8n_wf['id']}", headers=headers) as resp:
                if resp.status in (200, 204):
                    return {"status": "success", "workflowId": workflow_id, "message": f"{wf_info['name']} has been removed"}
                else:
                    error_text = await resp.text()
                    raise HTTPException(status_code=resp.status, detail=f"n8n API error: {error_text}")
    except aiohttp.ClientError as e:
        raise HTTPException(status_code=503, detail=f"Cannot reach n8n: {e}")


@router.get("/api/workflows/{workflow_id}/executions")
async def workflow_executions(workflow_id: str, limit: int = 20, api_key: str = Depends(verify_api_key)):
    """Get recent executions for a workflow."""
    n8n_workflows = await get_n8n_workflows()
    catalog = load_workflow_catalog()
    wf_info = next((wf for wf in catalog.get("workflows", []) if wf["id"] == workflow_id), None)
    if not wf_info:
        raise HTTPException(status_code=404, detail=f"Workflow not found: {workflow_id}")

    n8n_wf = None
    wf_name_lower = wf_info["name"].lower()
    for wf in n8n_workflows:
        if wf_name_lower in wf.get("name", "").lower():
            n8n_wf = wf
            break
    if not n8n_wf:
        return {"executions": [], "message": "Workflow not installed"}

    try:
        headers = {}
        if N8N_API_KEY:
            headers["X-N8N-API-KEY"] = N8N_API_KEY
        async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=5)) as session:
            async with session.get(f"{N8N_URL}/api/v1/executions", headers=headers, params={"workflowId": n8n_wf["id"], "limit": limit}) as resp:
                if resp.status == 200:
                    data = await resp.json()
                    return {"workflowId": workflow_id, "n8nId": n8n_wf["id"], "executions": data.get("data", [])}
                else:
                    return {"executions": [], "error": "Failed to fetch executions"}
    except (aiohttp.ClientError, OSError, json.JSONDecodeError):
        logger.exception("Failed to fetch workflow executions")
        return {"executions": [], "error": "Failed to fetch executions"}

File: dream-server/extensions/services/dashboard-api/crates/dashboard-api/src/routes/workflows.rs
//! Workflow router — /api/workflows/* endpoints. Mirrors routers/workflows.py.

use axum::extract::{Path, State};
use axum::Json;
use serde_json::{json, Value};
use std::path::PathBuf;

use crate::state::AppState;

fn workflow_dir() -> PathBuf {
    if let Ok(dir) = std::env::var("WORKFLOW_DIR") {
        return PathBuf::from(dir);
    }
    let install_dir = std::env::var("DREAM_INSTALL_DIR")
        .unwrap_or_else(|_| shellexpand::tilde("~/dream-server").to_string());
    let canonical = PathBuf::from(&install_dir).join("config").join("n8n");
    if canonical.exists() {
        canonical
    } else {
        PathBuf::from(&install_dir).join("workflows")
    }
}

fn n8n_url(services: &std::collections::HashMap<String, dream_common::manifest::ServiceConfig>) -> String {
    if let Ok(url) = std::env::var("N8N_URL") {
        return url;
    }
    if let Some(cfg) = services.get("n8n") {
        return format!("http://{}:{}", cfg.host, cfg.port);
    }
    "http://n8n:5678".to_string()
}

/// GET /api/workflows — list available workflow templates
pub async fn list_workflows() -> Json<Value> {
    let catalog_file = workflow_dir().join("catalog.json");
    if !catalog_file.exists() {
        return Json(json!({"workflows": [], "categories": {}}));
    }
    match std::fs::read_to_string(&catalog_file) {
        Ok(text) => {
            let data: Value = serde_json::from_str(&text).unwrap_or(json!({"workflows": [], "categories": {}}));
            Json(data)
        }
        Err(_) => Json(json!({"workflows": [], "categories": {}})),
    }
}

/// GET /api/workflows/:id — get a specific workflow template
pub async fn get_workflow(Path(id): Path<String>) -> Json<Value> {
    let wf_dir = workflow_dir();
    // Look in templates subdirectory
    let template_path = wf_dir.join("templates").join(format!("{id}.json"));
    if template_path.exists() {
        if let Ok(text) = std::fs::read_to_string(&template_path) {
            if let Ok(data) = serde_json::from_str::<Value>(&text) {
                return Json(data);
            }
        }
    }
    Json(json!({"error": "Workflow not found"}))
}

/// GET /api/workflows/categories — workflow categories from catalog
pub async fn workflow_categories() -> Json<Value> {
    let catalog_file = workflow_dir().join("catalog.json");
    if !catalog_file.exists() {
        return Json(json!({"categories": {}}));
    }
    match std::fs::read_to_string(&catalog_file) {
        Ok(text) => {
            let data: Value = serde_json::from_str(&text).unwrap_or(json!({}));
            Json(json!({"categories": data["categories"]}))
        }
        Err(_) => Json(json!({"categories": {}})),
    }
}

/// GET /api/workflows/n8n/status — n8n availability check
pub async fn n8n_status(State(state): State<AppState>) -> Json<Value> {
    let url = n8n_url(&state.services);
    let available = state
        .http
        .get(format!("{url}/healthz"))
        .send()
        .await
        .map(|r| r.status().as_u16() < 500)
        .unwrap_or(false);
    Json(json!({"available": available, "url": url}))
}

/// POST /api/workflows/:id/enable — import a workflow template into n8n
pub async fn enable_workflow(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Json<Value> {
    let url = n8n_url(&state.services);
    let n8n_key = std::env::var("N8N_API_KEY").unwrap_or_default();

    // Read the workflow template
    let catalog_file = workflow_dir().join("catalog.json");
    let catalog: Value = std::fs::read_to_string(&catalog_file)
        .ok()
        .and_then(|t| serde_json::from_str(&t).ok())
        .unwrap_or(json!({"workflows": []}));

    let wf_info = catalog["workflows"]
        .as_array()
        .and_then(|wfs| wfs.iter().find(|w| w["id"].as_str() == Some(id.as_str())));

    let wf_info = match wf_info {
        Some(w) => w.clone(),
        None => return Json(json!({"error": format!("Workflow not found: {id}")})),
    };

    let wf_file = wf_info["file"].as_str().unwrap_or("");
    let template_path = workflow_dir().join(wf_file);
    let template: Value = match std::fs::read_to_string(&template_path) {
        Ok(text) => serde_json::from_str(&text).unwrap_or(json!({})),
        Err(_) => return Json(json!({"error": format!("Workflow file not found: {wf_file}")})),
    };

    let mut headers = reqwest::header::HeaderMap::new();
    if !n8n_key.is_empty() {
        if let Ok(val) = n8n_key.parse() {
            headers.insert("X-N8N-API-KEY", val);
        }
    }

    match state
        .http
        .post(format!("{url}/api/v1/workflows"))
        .headers(headers.clone())
        .json(&template)
        .send()
        .await
    {
        Ok(resp) if resp.status().is_success() => {
            let data: Value = resp.json().await.unwrap_or(json!({}));
            let n8n_id = data["data"]["id"].as_str().map(|s| s.to_string());
            let mut activated = false;
            if let Some(ref nid) = n8n_id {
                if let Ok(r) = state
                    .http
                    .patch(format!("{url}/api/v1/workflows/{nid}"))
                    .headers(headers)
                    .json(&json!({"active": true}))
                    .send()
                    .await
                {
                    activated = r.status().is_success();
                }
            }
            Json(json!({
                "status": "success",
                "workflowId": id,
                "n8nId": n8n_id,
                "activated": activated,
                "message": format!("{} is now active!", wf_info["name"].as_str().unwrap_or(&id)),
            }))
        }
        Ok(resp) => {
            let status = resp.status().as_u16();
            let text = resp.text().await.unwrap_or_default();
            Json(json!({"error": format!("n8n API error ({status}): {text}")}))
        }
        Err(e) => Json(json!({"error": format!("Cannot reach n8n: {e}")})),
    }
}

/// POST /api/workflows/:id/disable — remove a workflow from n8n
pub async fn disable_workflow(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Json<Value> {
    let url = n8n_url(&state.services);
    let n8n_key = std::env::var("N8N_API_KEY").unwrap_or_default();

    // Find the n8n workflow ID by matching name from catalog
    let catalog_file = workflow_dir().join("catalog.json");
    let catalog: Value = std::fs::read_to_string(&catalog_file)
        .ok()
        .and_then(|t| serde_json::from_str(&t).ok())
        .unwrap_or(json!({"workflows": []}));

    let wf_info = catalog["workflows"]
        .as_array()
        .and_then(|wfs| wfs.iter().find(|w| w["id"].as_str() == Some(id.as_str())));

    let wf_info = match wf_info {
        Some(w) => w.clone(),
        None => return Json(json!({"error": format!("Workflow not found: {id}")})),
    };

    let mut headers = reqwest::header::HeaderMap::new();
    if !n8n_key.is_empty() {
        if let Ok(val) = n8n_key.parse() {
            headers.insert("X-N8N-API-KEY", val);
        }
    }

    // Fetch n8n workflows to find the matching ID
    let n8n_workflows: Vec<Value> = match state
        .http
        .get(format!("{url}/api/v1/workflows"))
        .headers(headers.clone())
        .send()
        .await
    {
        Ok(resp) if resp.status().is_success() => {
            let data: Value = resp.json().await.unwrap_or(json!({}));
            data["data"].as_array().cloned().unwrap_or_default()
        }
        _ => return Json(json!({"error": "Cannot reach n8n"})),
    };

    let wf_name = wf_info["name"].as_str().unwrap_or("").to_lowercase();
    let n8n_wf = n8n_workflows
        .iter()
        .find(|w| {
            let name = w["name"].as_str().unwrap_or("").to_lowercase();
            wf_name.contains(&name) || name.contains(&wf_name)
        });

    let n8n_wf = match n8n_wf {
        Some(w) => w,
        None => return Json(json!({"error": "Workflow not installed in n8n"})),
    };

    let n8n_id = n8n_wf["id"].as_str().unwrap_or("");
    match state
        .http
        .delete(format!("{url}/api/v1/workflows/{n8n_id}"))
        .headers(headers)
        .send()
        .await
    {
        Ok(resp) if resp.status().is_success() => {
            Json(json!({
                "status": "success",
                "workflowId": id,
                "message": format!("{} has been removed", wf_info["name"].as_str().unwrap_or(&id)),
            }))
        }
        Ok(resp) => {
            let text = resp.text().await.unwrap_or_default();
            Json(json!({"error": format!("n8n API error: {text}")}))
        }
        Err(e) => Json(json!({"error": format!("Cannot reach n8n: {e}")})),
    }
}

/// GET /api/workflows/:id/executions — recent executions for a workflow
pub async fn workflow_executions(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Json<Value> {
    let url = n8n_url(&state.services);
    let n8n_key = std::env::var("N8N_API_KEY").unwrap_or_default();

    let catalog_file = workflow_dir().join("catalog.json");
    let catalog: Value = std::fs::read_to_string(&catalog_file)
        .ok()
        .and_then(|t| serde_json::from_str(&t).ok())
        .unwrap_or(json!({"workflows": []}));

    let wf_info = catalog["workflows"]
        .as_array()
        .and_then(|wfs| wfs.iter().find(|w| w["id"].as_str() == Some(id.as_str())));

    let wf_info = match wf_info {
        Some(w) => w.clone(),
        None => return Json(json!({"error": format!("Workflow not found: {id}")})),
    };

    let mut headers = reqwest::header::HeaderMap::new();
    if !n8n_key.is_empty() {
        if let Ok(val) = n8n_key.parse() {
            headers.insert("X-N8N-API-KEY", val);
        }
    }

    // Fetch n8n workflows to find the matching ID
    let n8n_workflows: Vec<Value> = match state
        .http
        .get(format!("{url}/api/v1/workflows"))
        .headers(headers.clone())
        .send()
        .await
    {
        Ok(resp) if resp.status().is_success() => {
            let data: Value = resp.json().await.unwrap_or(json!({}));
            data["data"].as_array().cloned().unwrap_or_default()
        }
        _ => return Json(json!({"executions": [], "error": "Cannot reach n8n"})),
    };

    let wf_name = wf_info["name"].as_str().unwrap_or("").to_lowercase();
    let n8n_wf = n8n_workflows
        .iter()
        .find(|w| {
            let name = w["name"].as_str().unwrap_or("").to_lowercase();
            wf_name.contains(&name) || name.contains(&wf_name)
        });

    let n8n_wf = match n8n_wf {
        Some(w) => w,
        None => return Json(json!({"executions": [], "message": "Workflow not installed"})),
    };

    let n8n_id = n8n_wf["id"].as_str().unwrap_or("");
    match state
        .http
        .get(format!("{url}/api/v1/executions"))
        .headers(headers)
        .query(&[("workflowId", n8n_id), ("limit", "20")])
        .send()
        .await
    {
        Ok(resp) if resp.status().is_success() => {
            let data: Value = resp.json().await.unwrap_or(json!({}));
            Json(json!({
                "workflowId": id,
                "n8nId": n8n_id,
                "executions": data["data"],
            }))
        }
        _ => Json(json!({"executions": [], "error": "Failed to fetch executions"})),
    }
}

#[cfg(test)]
mod tests {
    use crate::state::AppState;
    use axum::body::Body;
    use dream_common::manifest::ServiceConfig;
    use http::Request;
    use http_body_util::BodyExt;
    use serde_json::{json, Value};
    use std::collections::HashMap;
    use tower::ServiceExt;
    use wiremock::matchers::{method, path};
    use wiremock::{Mock, MockServer, ResponseTemplate};

    const TEST_API_KEY: &str = "test-key-123";

    fn test_state_with_services(
        services: HashMap<String, ServiceConfig>,
    ) -> AppState {
        AppState::new(services, Vec::new(), Vec::new(), TEST_API_KEY.to_string())
    }

    fn test_state() -> AppState {
        test_state_with_services(HashMap::new())
    }

    fn app() -> axum::Router {
        crate::build_router(test_state())
    }

    fn app_with_services(services: HashMap<String, ServiceConfig>) -> axum::Router {
        crate::build_router(test_state_with_services(services))
    }

    fn auth_header() -> String {
        format!("Bearer {TEST_API_KEY}")
    }

    async fn get_auth(uri: &str) -> (http::StatusCode, Value) {
        let app = app();
        let req = Request::builder()
            .uri(uri)
            .header("authorization", auth_header())
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        let status = resp.status();
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap_or(json!(null));
        (status, val)
    }

    async fn get_auth_with_app(
        router: axum::Router,
        uri: &str,
    ) -> (http::StatusCode, Value) {
        let req = Request::builder()
            .uri(uri)
            .header("authorization", auth_header())
            .body(Body::empty())
            .unwrap();
        let resp = router.oneshot(req).await.unwrap();
        let status = resp.status();
        let body = resp.into_body().collect().await.unwrap().to_bytes();
        let val: Value = serde_json::from_slice(&body).unwrap_or(json!(null));
        (status, val)
    }

    // -----------------------------------------------------------------------
    // GET /api/workflows — no catalog file → empty response
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn list_workflows_returns_empty_when_no_catalog() {
        let tmp = tempfile::tempdir().unwrap();
        // Point WORKFLOW_DIR at an empty temp directory so no catalog.json exists.
        std::env::set_var("WORKFLOW_DIR", tmp.path().as_os_str());
        let (status, data) = get_auth("/api/workflows").await;
        std::env::remove_var("WORKFLOW_DIR");

        assert_eq!(status, http::StatusCode::OK);
        assert!(
            data.get("workflows").is_some(),
            "Expected 'workflows' key in response, got: {data}"
        );
        assert!(
            data["workflows"].as_array().unwrap().is_empty(),
            "Expected empty workflows array when catalog is missing, got: {data}"
        );
    }

    #[tokio::test]
    async fn list_workflows_requires_auth() {
        let app = app();
        let req = Request::builder()
            .uri("/api/workflows")
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), http::StatusCode::UNAUTHORIZED);
    }

    // -----------------------------------------------------------------------
    // GET /api/workflows/categories — no catalog → empty categories
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn workflow_categories_returns_categories_key() {
        let tmp = tempfile::tempdir().unwrap();
        std::env::set_var("WORKFLOW_DIR", tmp.path().as_os_str());
        let (status, data) = get_auth("/api/workflows/categories").await;
        std::env::remove_var("WORKFLOW_DIR");

        assert_eq!(status, http::StatusCode::OK);
        assert!(
            data.get("categories").is_some(),
            "Expected 'categories' key in response, got: {data}"
        );
    }

    // -----------------------------------------------------------------------
    // GET /api/workflows/categories — with catalog file
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn workflow_categories_reads_from_catalog() {
        let tmp = tempfile::tempdir().unwrap();
        let catalog = json!({
            "workflows": [],
            "categories": {
                "automation": "Automation",
                "ai": "AI & ML"
            }
        });
        std::fs::write(
            tmp.path().join("catalog.json"),
            serde_json::to_string(&catalog).unwrap(),
        )
        .unwrap();

        // Point WORKFLOW_DIR at the temp directory
        std::env::set_var("WORKFLOW_DIR", tmp.path().as_os_str());
        let (status, data) = get_auth("/api/workflows/categories").await;
        std::env::remove_var("WORKFLOW_DIR");

        assert_eq!(status, http::StatusCode::OK);
        let cats = &data["categories"];
        assert_eq!(cats["automation"], "Automation");
        assert_eq!(cats["ai"], "AI & ML");
    }

    // -----------------------------------------------------------------------
    // GET /api/workflows/{id} — workflow not found
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn get_workflow_not_found() {
        let tmp = tempfile::tempdir().unwrap();
        std::env::set_var("WORKFLOW_DIR", tmp.path().as_os_str());
        let (status, data) = get_auth("/api/workflows/does-not-exist").await;
        std::env::remove_var("WORKFLOW_DIR");

        assert_eq!(status, http::StatusCode::OK);
        assert_eq!(data["error"], "Workflow not found");
    }

    // -----------------------------------------------------------------------
    // GET /api/workflows/n8n/status — n8n unreachable
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn n8n_status_returns_unavailable_when_unreachable() {
        // No n8n service in state and no N8N_URL env → defaults to
        // http://n8n:5678 which is not reachable in tests.
        let (status, data) = get_auth("/api/workflows/n8n/status").await;
        assert_eq!(status, http::StatusCode::OK);
        assert_eq!(data["available"], false);
    }

    // -----------------------------------------------------------------------
    // GET /api/workflows/n8n/status — with mock n8n
    // -----------------------------------------------------------------------

    // -----------------------------------------------------------------------
    // GET /api/workflows/{id} — with catalog + template
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn get_workflow_with_template_returns_data() {
        let tmp = tempfile::tempdir().unwrap();
        std::fs::create_dir_all(tmp.path().join("templates")).unwrap();
        let template = json!({"id": "doc-qa", "name": "Document Q&A", "nodes": []});
        std::fs::write(
            tmp.path().join("templates/doc-qa.json"),
            serde_json::to_string(&template).unwrap(),
        )
        .unwrap();

        std::env::set_var("WORKFLOW_DIR", tmp.path().as_os_str());
        let (status, data) = get_auth("/api/workflows/doc-qa").await;
        std::env::remove_var("WORKFLOW_DIR");

        assert_eq!(status, http::StatusCode::OK);
        assert_eq!(data["id"], "doc-qa");
        assert_eq!(data["name"], "Document Q&A");
    }

    // -----------------------------------------------------------------------
    // Auth tests for enable/disable/executions
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn enable_workflow_requires_auth() {
        let app = app();
        let req = Request::builder()
            .method("POST")
            .uri("/api/workflows/test-wf/enable")
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), http::StatusCode::UNAUTHORIZED);
    }

    #[tokio::test]
    async fn disable_workflow_requires_auth() {
        let app = app();
        let req = Request::builder()
            .method("POST")
            .uri("/api/workflows/test-wf/disable")
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), http::StatusCode::UNAUTHORIZED);
    }

    #[tokio::test]
    async fn workflow_executions_requires_auth() {
        let app = app();
        let req = Request::builder()
            .uri("/api/workflows/test-wf/executions")
            .body(Body::empty())
            .unwrap();
        let resp = app.oneshot(req).await.unwrap();
        assert_eq!(resp.status(), http::StatusCode::UNAUTHORIZED);
    }

    // -----------------------------------------------------------------------
    // GET /api/workflows — with catalog file returns workflows
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn list_workflows_with_catalog_returns_data() {
        let tmp = tempfile::tempdir().unwrap();
        let catalog = json!({
            "workflows": [{"id": "wf1", "name": "Workflow 1"}],
            "categories": {"auto": "Automation"}
        });
        std::fs::write(
            tmp.path().join("catalog.json"),
            serde_json::to_string(&catalog).unwrap(),
        )
        .unwrap();

        std::env::set_var("WORKFLOW_DIR", tmp.path().as_os_str());
        let (status, data) = get_auth("/api/workflows").await;
        std::env::remove_var("WORKFLOW_DIR");

        assert_eq!(status, http::StatusCode::OK);
        let wfs = data["workflows"].as_array().unwrap();
        assert_eq!(wfs.len(), 1);
        assert_eq!(wfs[0]["id"], "wf1");
    }

    // -----------------------------------------------------------------------
    // GET /api/workflows/n8n/status — with mock n8n
    // -----------------------------------------------------------------------

    #[tokio::test]
    async fn n8n_status_returns_available_with_mock() {
        let mock_server = MockServer::start().await;
        Mock::given(method("GET"))
            .and(path("/healthz"))
            .respond_with(ResponseTemplate::new(200).set_body_json(json!({"status": "ok"})))
            .mount(&mock_server)
            .await;

        let mut services = HashMap::new();
        services.insert(
            "n8n".to_string(),
            ServiceConfig {
                host: mock_server.address().ip().to_string(),
                port: mock_server.address().port(),
                external_port: 5678,
                health: "/healthz".into(),
                name: "n8n".into(),
                ui_path: "/".into(),
                service_type: None,
                health_port: None,
            },
        );

        let router = app_with_services(services);
        let (status, data) = get_auth_with_app(router, "/api/workflows/n8n/status").await;
        assert_eq!(status, http::StatusCode::OK);
        assert_eq!(data["available"], true);
    }
}
